### ISSUE TITLE
CRAN release

### ISSUE THREAD

**Issue by mmcgillycuddy** [glmmTMB/glmmTMB#1132]
Hi there, just submitting an issue as per emails. Would it be possible to release a new version of glmmTMB on CRAN with a citation included in the DESCRIPTION and probably in the covstruc vignette as well? If so, would you know a rough timeline? Cheers!

**Comment by bbolker** [comment]
@mebrooks I think we're pretty much ready to go with this. I ran reverse-dependency checks over the weekend and didn't see any problems. I've already bumped the version number and included everything I could think of (or remember by looking at the git log) in the NEWS file ... (I didn't add the new citation to the vignette, though ...) 

If you can try tests locally and on win-builder I think we should be good to go ...

Note from Maeve's earlier e-mail that we will need to include a note with our submission saying that the DOI for the new citation doesn't resolve yet because JSS is waiting for us to push this release (chicken and egg problem ...)

**Comment by mebrooks** [comment]
@mmcgillycuddy did the journal give you a deadline? Is it ok if we take a few days to clean things up, knowing that a few days usually turns into a week?

**Comment by mmcgillycuddy** [comment]
Yes, but I asked for an extension.  They would just like to know a rough timeline. I am also organising others bits, so no pressure!

**Comment by mebrooks** [comment]
I merged the 2 recent pull requests, then sent it to Winbuilder and am checking reverse dependencies on DTU's cluster. 

I don't think I have time this week to get into testing priors with `tmbstan` so I'd consider the improvements @bbolker made to the MCMC vignette good enough for this release (but I won't close #1130 yet).

**Comment by bbolker** [comment]
I think we're ready to go if nothing goes wrong with RD checks.  The only other thing I can think of is including the JSS paper doi in the covstruct vignette (suggested by CRAN - we have it in the CITATION file (6c57f72131) but that's all.  (Making this change would *not* require redoing the RD checks.)

**Comment by mebrooks** [comment]
The latest version's Winbuilder checks were all fine. I'm doing RD checks again now. 

Does anyone want to add the citation to the covstruct vignette?

**Comment by mebrooks** [comment]
RD checks are still fine.

**Comment by mmcgillycuddy** [comment]
Oh I was just about to do it, but looks like @bbolker beat me to it.

**Comment by mebrooks** [comment]
OK, I think I'm rady to submit to CRAN. I confirmed that the citation shows up in the vignette. When I do `R CMD check --as-cran glmmTMB_1.1.11.tar.gz`, I get the expected notes about the DOI which I will mention in the submission. I also still get the same warning about Eigen that I had in #1084 and we can't do anything about.

**Comment by mebrooks** [comment]
CRAN says 
> Size of tarball: 6648015 bytes
> 
> Pls reduce to less than 5 MB for a CRAN package or explain why this increased that much.

I probably can't spend time on this until Monday.

**Comment by bbolker** [comment]
The tarball of the current version is 5.2 MB (so, already over the limit).  We added a few PNGs for the MCMC vignette (to avoid needing to do slow computations or rely on large stored objects), but those account for only a few 100K.  Not sure what else got bigger (attempt at a comparison below):

```
$ fdfind --newer 90days --size +100K | xargs ls -lSh
-rw-rw-r-- 1 bolker bolker 1.8M Sep 30 13:43 inst/test_data/models.rda
-rw-rw-r-- 1 bolker bolker 740K Sep 30 13:43 inst/vignette_data/troubleshooting.rda
-rw-rw-r-- 1 bolker bolker 737K Sep 30 13:43 inst/vignette_data/sherman.rda
-rw-rw-r-- 1 bolker bolker 705K Dec  3 19:47 vignettes/mcmc.html
-rw-rw-r-- 1 bolker bolker 308K Dec  9 11:14 vignettes/covstruct_cache/html/fit_spatial_model_3c485f7a9b6a2162c496157169886f62.rdb
-rw-rw-r-- 1 bolker bolker 285K Dec  9 11:26 vignettes/covstruct.html
-rw-rw-r-- 1 bolker bolker 269K Sep 30 13:43 inst/vignette_data/model_evaluation.rda
-rw-rw-r-- 1 bolker bolker 209K Dec  9 09:02 inst/vignette_data/mcmc.rda
-rw-rw-r-- 1 bolker bolker 103K Dec  9 09:02 inst/vignette_data/tmbstan_pairsplot.png
-rw-rw-r-- 1 bolker bolker  99K Nov 30 10:40 R/glmmTMB.R
```

**Comment by mmcgillycuddy** [comment]
@mebrooks @bbolker, it looks like I have missed the schedule for this volume. So sorry, but can we postpone the CRAN release? Or if you still want to do it, then the doi will need to be removed. I will get a new doi early next year for the next volume.

**Comment by mebrooks** [comment]
I don't mind waiting until early next year. Our last release was 2024-09-26, so it is a bit soon.

**Comment by bbolker** [comment]
Here are the largest new files with size differences. (see `misc/compare_tarball.R`) **tl;dr**  My tarball comes out to 6435135 bytes (slightly smaller than CRAN's). In any case, it looks like we might be including a couple of built HTML files that we don't need (`model_evaluation.html` in the vignettes dir seems like a mistake; `inst/doc/mcmc.html` is something we probably meant to do at some point, but maybe we don't need it?).

  You should try the `misc/compare_tarball.R` on the tarball you uploaded and see what you get (you'll need to update the file paths at the head of the file ...

``` 
                                                path size.old size.new
23                        glmmTMB/inst/doc/mcmc.html    21678   952972
232          glmmTMB/vignettes/model_evaluation.html       NA   636243
29             glmmTMB/inst/doc/model_evaluation.pdf   446235   445357
85   glmmTMB/inst/vignette_data/model_evaluation.rda   322075   322096
82               glmmTMB/inst/vignette_data/mcmc.rda    38525   213616
38                         glmmTMB/inst/doc/sim.html   247171   179750
17                      glmmTMB/inst/doc/glmmTMB.pdf   136136   136269
14                   glmmTMB/inst/doc/covstruct.html   119059   126438
94  glmmTMB/inst/vignette_data/tmbstan_pairsplot.png       NA   105354
161                              glmmTMB/R/glmmTMB.R   100786   101083
```

**Comment by bbolker** [comment]
Revisiting this:

* `glmmTMB/vignettes/model_evaluation.html` was just a mistake on my part (junk that got into the tarball by accident).
* The main difference is (I think) the increase in size of the MCMC vignette, because *we were (accidentally) not including any figures in the CRAN version of the vignette before* (which made it much less useful).

Hopefully if you go ahead and explain that they'll accept the current submission ...

**Comment by mebrooks** [comment]
I wrote to CRAN last night to explain the size. I didn't resubmit the package becasue they didn't say to do that in their email last week. I'll update here when I get a reply.

**Comment by strengejacke** [comment]
There are several ways to reduce the file size. One would be to render vignettes only online, and not include them in the package. Instead, just provide an "overview" (see https://cran.r-project.org/web/packages/parameters/index.html and full vignettes here: https://easystats.github.io/parameters/).

Another thing could be to put large models/data files on GitHub and download them. We have a dedicated repository for this (https://easystats.github.io/circus/), and in tests, we download them when an internet connection is available, e.g.

```r
test_that("icc", {
  skip_if_not_installed("curl")
  skip_if_offline()
  skip_if_not_installed("httr2")
  m2 <- insight::download_model("stanreg_lmerMod_1")
  expect_equal(
    icc(m2),
    data.frame(
      ICC_adjusted = 0.40579, ICC_conditional = 0.21881,
      ICC_unadjusted = 0.21881
    ),
    tolerance = 1e-2,
    ignore_attr = TRUE
  )
})
```

**Comment by strengejacke** [comment]
(putting model-files online also has the advantage that tests take less time for fitting models, addressing another CRAN requirement: CRAN checks should run in under x minutes)

**Comment by bbolker** [comment]
ideally I'd like CRAN to just agree to host the slightly larger package. (1) I like the package to be as self-contained as possible; (2) the larger size is precisely because we're already trying to minimize time spent on building vignettes.  Also, I know CRAN time is more valuable than ours, individually, but we already spend a lot of time jumping through CRAN hoops ...

**Comment by mebrooks** [comment]
As I got no reply and `foghorn` seems to say that we aren't in a queue, I'll try resubmitting tomorrow and explain the size change.

**Comment by bbolker** [comment]
Maybe since it's already January we should try to resynchronize with @mmcgillycuddy:s JSS paper?

**Comment by bbolker** [comment]
Heads up, @mmcgillycuddy is working on the final round of minor changes before submitting the final version of the JSS paper.  @mmcgillycuddy , can we have the new DOI please?

**Comment by bbolker** [comment]
bump for @mebrooks and @mmcgillycuddy : can we get the next release out?  (PRs #1147, #1148, #1152 are ready to go and should be harmless, although NEWS may need to be updated; #1153 would be nice but is probably too raw ...)

**Comment by mmcgillycuddy** [comment]
I have asked for a new doi, and hoping I will get it soon.

**Comment by bbolker** [comment]
Any update/worth nagging JSS for a new DOI?

**Comment by mmcgillycuddy** [comment]
No update, I'll check in with them.

**Comment by bbolker** [comment]
Any news from JSS about a doi ... ???

**Comment by bbolker** [comment]
@mebrooks  I think #1167, #1164, #1159, #1148 are all ready to go. #1154 is ready but fairly trivial (I just didn't want to lose that bit of utility coding).

**Comment by mmcgillycuddy** [comment]
@bbolker I asked when the doi will be released. I will let you know when they reply.

**Comment by mmcgillycuddy** [comment]
We will get a new doi this week

**Comment by bbolker** [comment]
For what it's worth we are now seeing test failures on CRAN, which are not our fault: they come ultimately from `nloptr`, see https://github.com/astamm/nloptr/issues/177#issuecomment-2707891429 . There's also a note about the use of `ENCLOS`, which is an upstream TMB issue, and NOTEs about the use of `\(x)`, which we fixed already.

**Comment by bbolker** [comment]
@mmcgillycuddy , any news on the new DOI ?  (The `nloptr` failures are resolved now ...)

**Comment by bbolker** [comment]
Sorry to nag ... (we'll shortly be seeing a warning from a test failure from the new version of `emmeans`; this is fixed by https://github.com/glmmTMB/glmmTMB/pull/1174 but we'll need to get it on CRAN soon-ish ...)

**Comment by mmcgillycuddy** [comment]
I have a doi. I will update the citation

**Comment by bbolker** [comment]
I see no reason to keep the arxiv one (I haven't checked your PR, but I assume that this is what you did ...)

I think we should merge #1167, #1173, then do winbuilder/reverse-dependency checks, then submit to CRAN! @mebrooks, do you agree?

**Comment by mebrooks** [comment]
Yep, sounds great! I have time over the next few days/ week.

**Comment by mmcgillycuddy** [comment]
Amazing. Thanks so much.

@bbolker yes, that's what I did. I didn't add it to the description because I am not sure how you would like to include it.

**Comment by bbolker** [comment]
I don't see why it needs to be in the Description: field of the DESCRIPTION file, we don't even have the 'primary' glmmTMB paper there ... but we should probably add it to the references in the covstruct vignette ...

**Comment by mebrooks** [comment]
I checked the reverse dependencies (now 55) and there are no new problems. Yay!

**Comment by strengejacke** [comment]
Just curious, how to you check reverse dependencies? Both the revdep package and base/tools R are making troubled every now and then, recently, this seems to be the most stable approach, though:

```r
tools::check_packages_in_dir(
  dir = "c:/Users/Daniel/Documents/GitHub/revdep/",
  check_args = c("--as-cran", ""),
  reverse = list(repos = getOption("repos")["CRAN"])
)
```

**Comment by bbolker** [comment]
I use `revdep` for `lme4`, but for whatever reason it's not very reliable - I catch some things, but miss a lot of stuff that CRAN subsequently finds ...

`revdep` has the nice feature that, in principle, it tests the dependencies against both the CRAN and the development versions of the package, allowing you to skip over (again, in principle) pre-existing problems that weren't your package's fault ...

**Comment by strengejacke** [comment]
It was somewhat "retired", but now this message was removed:

> This package is not actively maintained

Seems like revdep is being developed further again. But I also found revdep to be unreliable in the past, maybe it's getting better now that it's being worked on again.

**Comment by bbolker** [comment]
Again, if we can merge #1167, #1173 before release that would be nice (I don't expect downstream breakage from these, but to be properly paranoid we would rerun revdep checks after that ...)

**Comment by mebrooks** [comment]
OK, those 2 are merged. Can I submit it today, or are we waiting on anything else?

**Comment by bbolker** [comment]
Go for it

**Comment by mebrooks** [comment]
In the new round of `revdepcheck` results, it seems that #1167  may have broken `insight` and `sdmTMB`.

## insight
*   checking tests ...
   ```
      Running ‘testthat.R’
     ERROR
    Running the tests in ‘tests/testthat.R’ failed.
    Last 13 lines of output:
           ▆
        1. └─glmmTMB::glmmTMB(...) at test-null_model.R:25:3
        2.   ├─glmmTMB:::mkTMBStruc(...)
        3.   │ └─control$rank_check %in% c("warning", "stop", "adjust")
        4.   └─glmmTMB::glmmTMBControl()
        5.     ├─base::getOption("glmmTMB.autopar", get_autopar())
        6.     │ └─.Internal(getOption(x)) %||% default
        7.     └─glmmTMB:::get_autopar()
        8.       └─TMB::openmp()
        9.         ├─TMB::config(DLL = DLL)
       10.         └─TMB:::getUserDLL()
      
      [ FAIL 4 | WARN 1 | SKIP 73 | PASS 3452 ]
      Error: Test failures
      Execution halted
   ```

## sdmTMB
*   checking tests ...
    ```
      Running ‘testthat.R’
     ERROR
    Running the tests in ‘tests/testthat.R’ failed.
    Last 13 lines of output:
           ▆
        1. └─glmmTMB::glmmTMB(prop ~ 1, data = dat, weights = w, family = binomial()) at test-5-residuals.R:269:3
        2.   ├─glmmTMB:::mkTMBStruc(...)
        3.   │ └─control$rank_check %in% c("warning", "stop", "adjust")
        4.   └─glmmTMB::glmmTMBControl()
        5.     ├─base::getOption("glmmTMB.autopar", get_autopar())
        6.     │ └─.Internal(getOption(x)) %||% default
        7.     └─glmmTMB:::get_autopar()
        8.       └─TMB::openmp()
        9.         ├─TMB::config(DLL = DLL)
       10.         └─TMB:::getUserDLL()
      
      [ FAIL 1 | WARN 0 | SKIP 159 | PASS 185 ]
      Error: Test failures
      Execution halted
      ```


A week ago, I was warned by CRAN that we have to have a new version online by April 3rd. I didn't mention it before because I'm sure we can do that.

**Comment by seananderson** [comment]
For sdmTMB, at least, the issue seems to be that if sdmTMB is already loaded, then any call to glmmTMB (installed from the current GitHub main branch) returns:

```
Error in getUserDLL() :
    1   Multiple TMB models loaded. Failed to guess DLL name.
```

This does not happen prior to the `improve_par_control` pull request. For example,

```r
devtools::install_github("glmmTMB/glmmTMB/glmmTMB@79f4db2eb")
library(sdmTMB)
w <- sample(1:9, size = 300, replace = TRUE)
dat <- data.frame(y = stats::rbinom(300, size = w, 0.5))
dat$prop <- dat$y / w
m1 <- glmmTMB::glmmTMB(prop ~ 1, data = dat, weights = w, family = binomial())
```

is fine.

Replacing

```r
m1 <- glmmTMB::glmmTMB(prop ~ 1, data = dat, weights = w, family = binomial())
```

with

```r
library(glmmTMB)
m1 <- glmmTMB(prop ~ 1, data = dat, weights = w, family = binomial())
```

has the same behaviour.

**Comment by strengejacke** [comment]
From which branch can I install to test for insight?

**Comment by strengejacke** [comment]
I'm not sure this is an _insight_ issue, but rather an issue with _glmmTMB_?

This is what GitHub checks show:

```
══ Failed tests ════════════════════════════════════════════════════════════════
── Error ('test-model_data.R:27:1'): (code run outside of `test_that()`) ───────
Error in `getUserDLL()`: Multiple TMB models loaded. Failed to guess DLL name.
Backtrace:
     ▆
  1. └─glmmTMB::glmmTMB(...) at test-model_data.R:27:1
  2.   ├─glmmTMB:::mkTMBStruc(...) at glmmTMB/R/glmmTMB.R:1420:5
  3.   │ └─control$rank_check %in% c("warning", "stop", "adjust") at glmmTMB/R/glmmTMB.R:366:3
  4.   └─glmmTMB::glmmTMBControl() at glmmTMB/R/glmmTMB.R:1420:5
  5.     ├─base::getOption("glmmTMB.autopar", get_autopar()) at glmmTMB/R/glmmTMB.R:1530:5
  6.     │ └─.Internal(getOption(x)) %||% default
  7.     └─glmmTMB:::get_autopar()
  8.       └─TMB::openmp() at glmmTMB/R/utils.R:1044:5
  9.         ├─TMB::config(DLL = DLL)
 10.         └─TMB:::getUserDLL()
 11. ```
 12.

**Comment by bbolker** [comment]
Sorry, I'm just catching up now. I will look at this stuff and see if I can say anything useful ...

**Comment by bbolker** [comment]
My best guess after a very quick skim of [changes made in #1167](https://github.com/glmmTMB/glmmTMB/pull/1167/files) is that this happens (for some reason) because we're making an additional call to `TMB::openmp()` (this is also strongly suggested by the traceback shown above).  I think I might be able to fix this by adding `DLL = "glmmTMB"` to this call ...

**Comment by strengejacke** [comment]
The reprex doesn't fail. Not sure if it's related if some other glmmTMB model was tested before, and still is in the environment?

``` r
library(testthat)
skip_if_not_installed("glmmTMB")
data(Salamanders, package = "glmmTMB")
skip_on_os("mac") # error: FreeADFunObject
m <- glmmTMB::glmmTMB(
  count ~ spp + cover + mined + poly(DOP, 3) + (1 | site),
  ziformula = ~ spp + mined,
  dispformula = ~DOY,
  data = Salamanders,
  family = glmmTMB::nbinom2
)

model <- glmmTMB::glmmTMB(vs ~ disp, data = mtcars, family = binomial())
```

<sup>Created on 2025-03-28 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by strengejacke** [comment]
Testing the PR now.

**Comment by bbolker** [comment]
Ugh, stupid typo. Will fix ...

**Comment by bbolker** [comment]
Hmm.  I'm seeing some test failures locally with `insight`, but apparently unrelated to glmmTMB. Will be interested to see if the (very latest patched) PR resolves the current set of problems ...

**Comment by strengejacke** [comment]
Ok, will install from that PR again and re-check

**Comment by seananderson** [comment]
Yep, that `dll_check_fix` fixes it. This:

```r
devtools::install_github("glmmTMB/glmmTMB/glmmTMB@dll_check_fix")
library(sdmTMB)
w <- sample(1:9, size = 300, replace = TRUE)
dat <- data.frame(y = stats::rbinom(300, size = w, 0.5))
dat$prop <- dat$y / w
m1 <- glmmTMB::glmmTMB(prop ~ 1, data = dat, weights = w, family = binomial())
```

runs without issue.

**Comment by bbolker** [comment]
@mebrooks I think we should be ready to go now ... ?

**Comment by mebrooks** [comment]
I submitted it yesterday and got this reply.

> Size of tarball: 6870410 bytes
> 
> Can you please keep this below the 5MB threshold?

**Comment by mebrooks** [comment]
I think there's potential to reduce the size of files associated with the troubleshooting vignette. I don't think it ever uses `troubleshooting.rda` which is 756959 bytes so we could delete it along with `troubleshooting.R`.  Looking at `glmmTMB/inst/vignette_data/troubleshooting.R` and `glmmTMB/inst/vignette_data/sherman.R`, they seem to be redundant except for `mod4` which doen't get used. Also, I think we can remove `dd` from `sherman.rda` as it doesn't get used. 

Do we need to include `framed.sty` and `lineno.sty`? [I see framed.sty is used by `knitr`](https://github.com/yihui/knitr/issues/598#issuecomment-46700382).  The model evalutaion vignette includes `\usepackage{lineno}`, but does it actually get used?

**Comment by bbolker** [comment]
Sigh. 

The troubleshooting vignette needs some love. Note that the [rendered version on CRAN](https://cran.r-project.org/web/packages/glmmTMB/vignettes/troubleshooting.html) doesn't have any of the outputs. The point of `troubleshooting.rda` was to be able to load stored fits, so that we could have a rendered version with the outputs but not have CRAN yell at us because the vignettes took too long to render. Maybe we should just move to [including static vignettes with the package](https://stackoverflow.com/a/24111191/190277) ?

I would strongly prefer to put the unused stuff in `.Rbuildignore` for now rather than deleting the files; I know we can always retrieve them from the git history, but I'd rather have a little bit of chaff in the repo (ideally with a README file documenting what we did) then have to do a lot of git archaeology ...

I guess we could probably do without `lineno.sty`.  Worrying about individual files that are only 150K feels like a big waste of time to me ... (You're right that `\linenumbers` is commented out in the vignette right now.)

**Comment by strengejacke** [comment]
Upload the models for testing to the [circus-repo](https://easystats.github.io/circus/) and use `insight::download_model()` in tests, saves a lot of space ;-)

**Comment by mebrooks** [comment]
I'm open to other solutions. The changes in #1183 were just the simplest things I could find. Although, if we go with that option, we still need to get rid of 0.3 MB.

**Comment by bbolker** [comment]
Working on a branch that will trim the traceplots from the MCMC vignette (and `.Rbuildignore` `troubleshooting.rmd` for now.

@strengejacke, I appreciate the solution, but I also feel strongly about keeping the package self-contained if at all possible (maybe it never happens any more, but I've been in workshop situations with bad or nonexistent network access where not being able to see good documentation would have been really frustrating).

**Comment by strengejacke** [comment]
Yeah, I don't mean all the stuff, only models/ data that is not user-visible and solely intended for testing. And it doesn't need to be our repo :-)

**Comment by bbolker** [comment]
I have started hacking troubleshooting stuff to take less room but I'm not there yet. Sigh. Will work on this more later. https://github.com/glmmTMB/glmmTMB/tree/squash_mcmc_vignette

**Comment by bbolker** [comment]
https://github.com/glmmTMB/glmmTMB/pull/1184 . It might be worth taking a quick look to make sure I haven't f*ed up the MCMC and troubleshooting vignettes, but it passes `R CMD check --as-cran` on my machine, the vignettes should build OK on CRAN as well, and the tarball is <5M ...

**Comment by strengejacke** [comment]
> For the future I definitely think that downloading some externally stored models for testing purposes might be a good idea (would be nice if we can do it with base-R tools so that we don't further drag in the whole tidyverse stack ...)

Would also be interested in a base-R solution (although you don't need to put all packages in _Imports_, but rather in _Suggests_).

**Comment by bbolker** [comment]
Haven't figured it out yet, but I pushed a minimal version of this to `misc/qra_debug.R` (saved the problematic object created in the `qra` vignette. Something deep in `vcov()`. Possibly from 6bc5ca236c31b1f ?

**Comment by bbolker** [comment]
Aha. I think I see the problem. The vignette is fitting a binomial model with a dispformula, but binomial models don't have dispersion components!

I think we should probably ask the maintainer (Maindonald) to change the vignette, and add an error when someone specifies a dispersion model for a non-dispersion family (binomial, Poisson ...)

**Comment by bbolker** [comment]
https://github.com/glmmTMB/glmmTMB/pull/1185

**Comment by mebrooks** [comment]
This slipped by because `qra` failed to install automatically via `revdepcheck` on the cluster where I did the checks. It also happened with `ecostats`. They both install OK via the normal way on the cluster (launch R from a Terminal and then `install.packages()`), so I wonder what `revdepcheck` is doing differently.

```
Error: package or namespace load failed for ‘latticeExtra’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/zhome/61/6/119429/glmmTMB/glmmTMB/revdep/library/qra/png/libs/png.so':
  libpng15.so.15: cannot open shared object file: No such file or directory
Execution halted
ERROR: lazy loading failed for package ‘qra’
```

```
Error: package or namespace load failed for ‘mvabund’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/zhome/61/6/119429/glmmTMB/glmmTMB/revdep/library/ecostats/mvabund/libs/mvabund.so':
  libgsl.so.0: cannot open shared object file: No such file or directory
Execution halted
ERROR: lazy loading failed for package ‘ecostats’
```

**Comment by mebrooks** [comment]
Success!

**Comment by bbolker** [comment]
Since I can only leave one reaction above: :tada: :tada: :tada: :tada:

**Comment by mmcgillycuddy** [comment]
Amazing. Thank you. What a mission!
